app_dart/lib/src/service/github_status_service.dart

```diff
@@ -0,0 +1,109 @@
+// Copyright 2020 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:github/github.dart';
+import 'package:meta/meta.dart';
+
+import '../datastore/cocoon_config.dart';
+import '../model/luci/buildbucket.dart' as bb;
+import '../model/luci/push_message.dart';
+import 'luci_build_service.dart';
+
+class GithubStatusService {
+  const GithubStatusService(this.config, this.luciBuildService);
+
+  /// The global configuration of this AppEngine server.
+  final Config config;
+  final LuciBuildService luciBuildService;
+
+  Future<void> setBuildsPendingStatus(
+      String repositoryName, int prNumber, String commitSha) async {
+    final RepositorySlug slug = RepositorySlug('flutter', repositoryName);
+    final GitHub gitHubClient = await config.createGitHubClient();
+    final Map<String, bb.Build> builds = await luciBuildService
+        .buildsForRepositoryAndPr(repositoryName, prNumber, commitSha);
+    for (bb.Build build in builds.values) {
+      final CreateStatus status = CreateStatus('pending')
+        ..context = build.builderId.builder
+        ..description = 'Flutter LUCI Build: ${build.builderId.builder}'
+        ..targetUrl = '';
+      await gitHubClient.repositories.createStatus(slug, commitSha, status);
+    }
+  }
+
+  Future<void> setPendingStatus({
+    @required String ref,
+    @required String builderName,
+    @required String buildUrl,
+  }) async {
+    final RepositorySlug slug = await config.repoNameForBuilder(builderName);
+    // No builderName configuration, nothing to do here.
+    if (slug == null) {
+      return;
+    }
+    final GitHub gitHubClient = await config.createGitHubClient();
+    // GitHub "only" allows setting a status for a context/ref pair 1000 times.
+    // We should avoid unnecessarily setting a pending status, e.g. if we get
+    // started and pending messages close together.
+    // We have to check for both because sometimes one or the other might come
+    // in.
+    // However, we should keep going if the _most recent_ status is not pending.
+    await for (RepositoryStatus status
+        in gitHubClient.repositories.listStatuses(slug, ref)) {
+      if (status.context == builderName) {
+        if (status.state == 'pending' &&
```

Comment:
Is this a good time to use a variable instead of 'pending' everywhere?

Reply:
Done